README.md
## Library to integrate [Livechat](https://www.livechat.com/) with Angular

[![npm version](https://badge.fury.io/js/%40alekremi%2Fngx-livechat.svg)](https://badge.fury.io/js/%40alekremi%2Fngx-livechat)
![npm bundle size](https://img.shields.io/bundlephobia/minzip/@alekremi/ngx-livechat)
![Depfu](https://img.shields.io/depfu/alekremi/ngx-livechat)

### Demo

StackBlitz [live example](https://stackblitz.com/edit/ngx-livechat-example).

### Installation

```bash
npm install --save @alekremi/ngx-livechat
```

### Usage

Import `NgxLivechatModule` in Angular AppModule.

```javascript
import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { NgxLivechatModule } from 'ngx-livechat';

import { AppComponent } from './app.component';

@NgModule({
  imports: [BrowserModule, NgxLivechatModule.forRoot({ license: 11082047 })],
  declarations: [AppComponent],
  bootstrap: [AppComponent]
})
export class AppModule {}
```

The `NgxLivechatService` also provides methods for LiveChat APIs, such as:

- `open_chat_window`
- `minimize_chat_window` and etc.

For example, how to programmatically open a LiveChat window:

```javascript
constructor(private livechatService: NgxLivechatService) {}

openChatWindow(): void {
    this.livechatService.openChatWindow();
}

```

### Advanced usage

Using external configuration

For example your external configuration looks like:

```javascript
interface ExternalConfig {
  liveChatLicense: number;
  //...any other params
}
```

First, we have to set manual initialization flag and specify any license number (for example, 0, this license number will be overwritten when the API service returns the external configuration).

```javascript
NgxLivechatModule.forRoot({ license: 0, initManually: true });
```

app.module.ts example:

```javascript
import { APP_INITIALIZER, NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { HttpClientModule, HttpClient } from '@angular/common/http';
import { tap } from 'rxjs/operators';
import { NgxLivechatModule, NgxLivechatConfig, NGX_LIVECHAT_CONFIG } from '@alekremi/ngx-livechat';

import { AppComponent } from './app.component';

export function liveChatConfigFactory(httpClient: HttpClient, config: NgxLivechatConfig) {
  return (): Promise<ExternalConfig> =>
    httpClient.get <
    ExternalConfig >
    'YOUR_API_URL'.pipe(tap((externalConfig) => (config.license = externalConfig.liveChatLicense))).toPromise();
}

@NgModule({
  imports: [BrowserModule, HttpClientModule, NgxLivechatModule.forRoot({ license: 0, initManually: true })],
  providers: [
    {
      provide: APP_INITIALIZER,
      useFactory: liveChatConfigFactory,
      deps: [HttpClient, NGX_LIVECHAT_CONFIG],
      multi: true
    }
  ],
  declarations: [AppComponent],
  bootstrap: [AppComponent]
})
export class AppModule {}
```

Then initialize LiveChat mannualy for example in app.component.ts:

```javascript
import { Component, OnInit } from '@angular/core';
import { NgxLivechatService } from '@alekremi/ngx-livechat';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css']
})
export class AppComponent implements OnInit {
  constructor(private livechatService: NgxLivechatService) {}

  ngOnInit(): void {
    this.livechatService.initLiveChat();
  }
}
```

### Configuration

| Parameter         | Optional | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                   |
| ----------------- | -------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| license           | No       | LiveChat license number.                                                                                                                                                                                                                                                                                                                                                                                                                                      |
| group             | Yes      | Chat window group (defaults to "0"). You can divide LiveChat agents into different departments, such as "Billing" or "Support". <br>For example, if this parameter will point to group "Billing", all visitors entering the chat will talk with agents from this group and not the "Support" group. groups                                                                                                                                                    |
| chatBetweenGroups | Yes      | By default, visitor's browsing history is remembered across different groups. <br><br> If you don't want to display visitor's browsing history across different groups, use the following code. <br><br>Using this parameter is not recommended when using target field in the pre-chat survey.                                                                                                                                                               |
| params            | Yes      | Custom variables sent to LiveChat applications.<br>These can be your visitor's account ID, login or any other information that is important for LiveChat agent during the chat. <br><br>"name" can be max 500 characters long.<br>"value" can be max 3500 characters long.                                                                                                                                                                                    |
| visitor           | Yes      | Visitor's data. If your visitor is already logged in to your system, you can pass his name and e-mail to LiveChat apps.<br>Agents will see the information on the "Visitors" list and in the Archives.                                                                                                                                                                                                                                                        |
| gaVersion         | Yes      | By default, our tracking code strores LiveChat related data in the Google Analytics gaq - traditional asynchronous code for Google Analytics.<br><br>If you are using a different type of Google Analytics, you can decide which one LiveChat should track.<br><br>The available options are: <br>ga – Universal Analytics;<br> gtm – Google Tag Manager;<br> gtag – Global Site Tag (gtag.js);<br> gaq – traditional asynchronous code for Google Analytics. |
| initManually      | Yes      | The parameter determines whether LiveChat will be manually initialized. By default, it is initialized when the application starts.                                                                                                                                                                                                                                                                                                                            |


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.1.0](https://github.com/alekremi/ngx-livechat/compare/v1.0.0...v1.1.0) (2021-07-06)

### Features

- added the feature to initialize livechat widget manually , updated the documentation (advanced usage) ([c2d5623](https://github.com/alekremi/ngx-livechat/commit/c2d5623266aa0e8830048ad765bc1807bf43fb6f))

## 1.0.0 (2021-07-05)

### Features

- livechat module and service implemented ([3dc5e2a](https://github.com/alekremi/ngx-livechat/commit/3dc5e2a9bfc3afa598c9962da0ec8666dd3599b1))


projects/ngx-livechat/src/lib/ngx-livechat.service.spec.ts
import { TestBed } from '@angular/core/testing';

import { NgxLivechatModule } from './ngx-livechat.module';
import { NgxLivechatService } from './ngx-livechat.service';

describe('NgxLivechatService', () => {
  let service: NgxLivechatService;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [NgxLivechatModule.forRoot({ license: 11082047 })]
    });
    service = TestBed.inject(NgxLivechatService);
  });

  it('should be created', () => {
    expect(service).toBeTruthy();
  });
});


projects/ngx-livechat/karma.conf.js
// Karma configuration file, see link for more information
// https://karma-runner.github.io/1.0/config/configuration-file.html

module.exports = function (config) {
  config.set({
    basePath: '',
    frameworks: ['jasmine', '@angular-devkit/build-angular'],
    plugins: [
      require('karma-jasmine'),
      require('karma-chrome-launcher'),
      require('karma-jasmine-html-reporter'),
      require('karma-coverage'),
      require('@angular-devkit/build-angular/plugins/karma')
    ],
    client: {
      jasmine: {
        // you can add configuration options for Jasmine here
        // the possible options are listed at https://jasmine.github.io/api/edge/Configuration.html
        // for example, you can disable the random execution with `random: false`
        // or set a specific seed with `seed: 4321`
      },
      clearContext: false // leave Jasmine Spec Runner output visible in browser
    },
    jasmineHtmlReporter: {
      suppressAll: true // removes the duplicated traces
    },
    coverageReporter: {
      dir: require('path').join(__dirname, '../../coverage/ngx-livechat'),
      subdir: '.',
      reporters: [
        { type: 'html' },
        { type: 'text-summary' }
      ]
    },
    reporters: ['progress', 'kjhtml'],
    port: 9876,
    colors: true,
    logLevel: config.LOG_INFO,
    autoWatch: true,
    browsers: ['Chrome'],
    singleRun: false,
    restartOnFileChange: true
  });
};


projects/ngx-livechat/src/lib/ngx-livechat.module.ts
import { APP_INITIALIZER, ModuleWithProviders, NgModule, Optional, SkipSelf } from '@angular/core';

import { NgxLivechatService } from './ngx-livechat.service';
import { NgxLivechatConfig, NGX_LIVECHAT_CONFIG } from './ngx-livechat';

@NgModule({})
export class NgxLivechatModule {
  constructor(@Optional() @SkipSelf() parentModule: NgxLivechatModule) {
    if (parentModule) {
      throw new Error('NgxLivechatModule is already loaded, import it only once.');
    }
  }

  static forRoot(config: NgxLivechatConfig): ModuleWithProviders<NgxLivechatModule> {
    const moduleWithProviders: ModuleWithProviders<NgxLivechatModule> = {
      ngModule: NgxLivechatModule,
      providers: [{ provide: NGX_LIVECHAT_CONFIG, useValue: config }]
    };
    if (!config.initManually) {
      moduleWithProviders.providers?.push({
        provide: APP_INITIALIZER,
        useFactory: (service: NgxLivechatService) => () => service.initLiveChat(),
        deps: [NgxLivechatService],
        multi: true
      });
    }
    return moduleWithProviders;
  }
}


projects/ngx-livechat/src/lib/ngx-livechat.service.ts
import { DOCUMENT } from '@angular/common';
import { Inject, Injectable } from '@angular/core';
import { Observable, Subject } from 'rxjs';

import { NgxLivechatConfig, NgxLivechatWindow, NGX_LIVECHAT_CONFIG } from './ngx-livechat';

@Injectable({ providedIn: 'root' })
export class NgxLivechatService {
  liveChatLoaded$: Observable<boolean>;

  private liveChatLoadedSubject: Subject<boolean>;
  private window: NgxLivechatWindow;

  constructor(@Inject(DOCUMENT) private document: Document, @Inject(NGX_LIVECHAT_CONFIG) private config: NgxLivechatConfig) {
    this.window = this.document.defaultView as unknown as NgxLivechatWindow;
    this.liveChatLoadedSubject = new Subject<boolean>();
    this.liveChatLoaded$ = this.liveChatLoadedSubject.asObservable();
  }

  initLiveChat(): void {
    if (!this.window.__lc) {
      this.window.__lc = {
        license: this.config.license,
        group: this.config.group,
        chat_between_groups: this.config.chatBetweenGroups,
        params: this.config.params,
        visitor: this.config.visitor,
        ga_version: this.config.gaVersion
      };

      const script = this.createHTMLScriptElement('https://cdn.livechatinc.com/tracking.js', true);
      this.document.body.insertBefore(script, this.document.body.firstChild);

      script.addEventListener('load', () => (this.window.LC_API.on_after_load = () => this.liveChatLoadedSubject.next(true)));
      script.addEventListener('error', (error) => console.error(error));
    }
  }

  agentsAreAvailable = (): boolean => this.window.LC_API.agents_are_available();
  chatRunning = (): boolean => this.window.LC_API.chat_running();
  chatWindowHidden = (): boolean => this.window.LC_API.chat_window_hidden();
  chatWindowMaximized = (): boolean => this.window.LC_API.chat_window_maximized();
  chatWindowMinimized = (): boolean => this.window.LC_API.chat_window_minimized();
  closeChat = (): void => this.window.LC_API.close_chat();
  getChatId = (): string => this.window.LC_API.get_chat_id();
  getChatsNumber = (): number => this.window.LC_API.get_chats_number();
  getLastVisitTimestamp = (): Date => this.window.LC_API.get_last_visit_timestamp();
  getPageViewsNumber = (): number => this.window.LC_API.get_page_views_number();
  getVisitorId = (): string => this.window.LC_API.get_visitor_id();
  getVisitsNumber = (): number => this.window.LC_API.get_visits_number();
  hideChatWindow = (): void => this.window.LC_API.hide_chat_window();
  minimizeChatWindow = (): void => this.window.LC_API.minimize_chat_window();
  openChatWindow = (): void => this.window.LC_API.open_chat_window();
  setCustomVariables = (variables: unknown): void => this.window.LC_API.set_custom_variables(variables);
  setVisitorEmail = (email: string): void => this.window.LC_API.set_visitor_email(email);
  setVisitorName = (name: string): void => this.window.LC_API.set_visitor_name(name);
  triggerSalesTracker = (e: unknown, t: unknown): void => this.window.LC_API.trigger_sales_tracker(e, t);
  updateCustomVariables = (variables: unknown): void => this.window.LC_API.update_custom_variables(variables);
  visitorEngaged = (): void => this.window.LC_API.visitor_engaged();
  visitorQueued = (): void => this.window.LC_API.visitor_queued();

  private createHTMLScriptElement(src: string, async?: boolean): HTMLScriptElement {
    const el: HTMLScriptElement = document.createElement('script');
    el.type = 'text/javascript';
    el.src = src;
    el.async = async || false;
    return el;
  }
}


projects/ngx-livechat/src/lib/ngx-livechat.ts
import { InjectionToken } from '@angular/core';

export interface NgxLivechatConfig {
  license: number;
  group?: number;
  chatBetweenGroups?: boolean;
  params?: { name: string; value: string }[];
  visitor?: { name: string; email: string };
  gaVersion?: 'ga' | 'gtm' | 'gtag' | 'gaq';
  initManually?: boolean;
}

export const NGX_LIVECHAT_CONFIG = new InjectionToken<NgxLivechatConfig>('NGX_LIVECHAT_CONFIG');

export interface NgxLiveChat {
  /**
   * LiveChat license number
   */
  license: number;

  /**
   * Chat window group (defaults to "0").
   * You can divide LiveChat agents into different departments,
   * such as "Billing" or "Support".
   * For example, if this parameter will point to group "Billing",
   * all visitors entering the chat will talk with agents
   * from this group and not the "Support" group.
   *
   * Create your group in LiveChat app:
   * https://my.livechatinc.com/agents/groups
   */
  group?: number;

  /**
   * By default, visitor's browsing history is remembered
   * across different groups.
   *
   * If you don't want to display visitor's browsing history
   * across different groups, use the following code.
   *
   * Using this parameter is not recommended when
   * using target field in the pre-chat survey.
   */
  chat_between_groups?: boolean;

  /**
   * Custom variables sent to LiveChat applications.
   * These can be your visitor's account ID, login
   * or any other information that is important for
   * LiveChat agent during the chat.
   *
   * "name"  can be max 500 characters long.
   * "value" can be max 3500 characters long.
   */
  params?: { name: string; value: string }[];

  /**
   * Visitor's data. If your visitor is already logged in
   * to your system, you can pass his name and e-mail to LiveChat apps.
   * Agents will see the information on the "Visitors" list
   * and in the Archives.
   */
  visitor?: { name: string; email: string };

  /**
   * By default, our tracking code strores LiveChat related data
   * in the Google Analytics gaq - traditional asynchronous
   * code for Google Analytics.
   *
   * If you are using a different type of Google Analytics,
   * you can decide which one LiveChat should track.
   *
   * The available options are:
   * ga – Universal Analytics;
   * gtm – Google Tag Manager;
   * gtag – Global Site Tag (gtag.js);
   * gaq – traditional asynchronous code for Google Analytics.
   */
  ga_version?: 'ga' | 'gtm' | 'gtag' | 'gaq';
}

export interface NgxLiveChatApi {
  agents_are_available(): boolean;
  chat_running(): boolean;
  chat_window_hidden(): boolean;
  chat_window_maximized(): boolean;
  chat_window_minimized(): boolean;
  close_chat(): void;
  get_chat_id(): string;
  get_chats_number(): number;
  get_last_visit_timestamp(): Date;
  get_page_views_number(): number;
  get_visitor_id(): string;
  get_visits_number(): number;
  hide_chat_window(): void;
  minimize_chat_window(): void;
  on_after_load(): void;
  on_before_load(): void;
  on_chat_ended(): void;
  on_chat_started(data: unknown): void;
  on_chat_state_changed(data: unknown): void;
  on_chat_window_hidden(): void;
  on_chat_window_minimized(): void;
  on_chat_window_opened(): void;
  on_message(data: unknown): void;
  on_postchat_survey_submitted(data: unknown): void;
  on_prechat_survey_submitted(data: unknown): void;
  on_rating_comment_submitted(data: unknown): void;
  on_rating_submitted(data: unknown): void;
  on_ticket_created(data: unknown): void;
  on_widget_resize(): void;
  open_chat_window(): void;
  set_custom_variables(variables: unknown): void;
  set_visitor_email(email: string): void;
  set_visitor_name(name: string): void;
  trigger_sales_tracker(e: unknown, t: unknown): void;
  update_custom_variables(variables: unknown): void;
  visitor_engaged(): void;
  visitor_queued(): void;
}

export interface NgxLivechatWindow extends Window {
  LiveChatWidget: unknown;
  LC_API: NgxLiveChatApi;
  __lc: NgxLiveChat;
}
